File: lib/src/main/java/com/tencent/cubershi/plugin_loader/classloaders/PluginClassLoader.java
package com.tencent.cubershi.plugin_loader.classloaders;

import com.tencent.cubershi.plugin_loader.exceptions.LoadApkException;

import java.io.File;

public class PluginClassLoader extends ClassLoader {
    /**
     * 用于返回自定义的Android组件的ClassLoader
     */
    final private MockBootClassLoader mMockBootClassLoader;
    /**
     * 用于返回插件类的ClassLoader
     */
    final private PluginDexClassLoader mPluginDexClassLoader;

    public PluginClassLoader(ClassLoader parent, MockBootClassLoader mockBootClassLoader, File apk, File odexDir, File libDir) throws LoadApkException {
        super(parent);
        this.mMockBootClassLoader = mockBootClassLoader;
        mPluginDexClassLoader = buildPluginDexClassLoader(parent, apk, odexDir, libDir);
    }

    private PluginDexClassLoader buildPluginDexClassLoader(ClassLoader parent, File apk, File odexDir, File libDir) throws LoadApkException {
        if (odexDir.exists() && !odexDir.isDirectory()) {
            throw new LoadApkException("odexDir目标路径" + odexDir.getAbsolutePath()
                    + "已被其他文件占用");
        } else if (!odexDir.exists()) {
            boolean success = odexDir.mkdir();
            if (!success) {
                throw new LoadApkException("odexDir目标路径" + odexDir.getAbsolutePath()
                        + "创建目录失败");
            }
        }

        if (!libDir.exists()) {
            if (!libDir.mkdirs()) {
                throw new LoadApkException("libDir目标路径" + libDir.getAbsolutePath()
                        + "创建目录失败");
            }
        }

        return new PluginDexClassLoader(apk.getAbsolutePath(), odexDir.getAbsolutePath(), libDir.getAbsolutePath(), parent);
    }

    @Override
    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        Class<?> c = findLoadedClass(name);
        if (c == null) {
            try {
                c = mMockBootClassLoader.findClass(name);
            } catch (ClassNotFoundException e) {
                c = mPluginDexClassLoader.loadClassX(name, resolve);
            }
        }
        return c;
    }
}

File: shadow-sdk/shadow-loader/src/main/kotlin/com/tencent/shadow/loader/classloaders/PluginClassLoader.kt
package com.tencent.shadow.loader.classloaders

import android.content.Context
import android.os.Build
import com.tencent.shadow.loader.classloaders.multidex.MultiDex
import dalvik.system.DexClassLoader
import java.io.File


/**
 * 用于加载插件的ClassLoader.
 *
 * @author cubershi
 */
class PluginClassLoader(
        hostAppContext: Context, dexPath: String, optimizedDirectory: String, private val librarySearchPath: String, parent: ClassLoader
) : DexClassLoader(dexPath, optimizedDirectory, librarySearchPath, parent) {

    private val mGrandParent: ClassLoader

    init {
        if (Build.VERSION.SDK_INT <= MultiDex.MAX_SUPPORTED_SDK_VERSION) {
            val pluginLoaderMultiDex = hostAppContext.getSharedPreferences("com.tencent.shadow.multidex", Context.MODE_PRIVATE)
            MultiDex.install(this, dexPath, File(optimizedDirectory), pluginLoaderMultiDex)
        }
        mGrandParent = parent.parent
    }

    fun getLibrarySearchPath() = librarySearchPath

    @Throws(ClassNotFoundException::class)
    override fun loadClass(className: String, resolve: Boolean): Class<*> {
        if (className.startsWith("com.tencent.shadow.runtime")
                || className.startsWith("org.apache.commons.logging")//org.apache.commons.logging是非常特殊的的包,由系统放到App的PathClassLoader中.
          || (Build.VERSION.SDK_INT < 28 && className.startsWith("org.apache.http"))) {//Android 9.0以下的系统里面带有http包，走系统的不走本地的
            return super.loadClass(className, resolve)
        } else {
            var clazz: Class<*>? = findLoadedClass(className)

            if (clazz == null) {
                var suppressed: ClassNotFoundException? = null
                try {
                    clazz = findClass(className)!!
                } catch (e: ClassNotFoundException) {
                    suppressed = e
                }

                if (clazz == null) {
                    try {
                        clazz = mGrandParent.loadClass(className)!!
                    } catch (e: ClassNotFoundException) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                            e.addSuppressed(suppressed)
                        }
                        throw e
                    }

                }
            }

            return clazz
        }
    }
}
